src/st/no0/57C20.c

```diff
@@ -51,7 +51,518 @@ void func_us_801D91C4(Entity* self) {
+                if (GetDistanceToPlayerX() > 0x40) {
+                    self->facingLeft = (GetSideToPlayer() & 1) ^ 1;
+                }
+                SetSubStep(0);
+
+                if (self->ext.et_801D9264.unk84 == 1) {
+                    SetStep(6);
+                    posX = self->posX.i.hi + g_Tilemap.scrollX.i.hi;
+                    if (posX > 0x400) {
+                        self->facingLeft = 1;
+                        self->ext.et_801D9264.unk85 = 3;
+                    }
+                    if (++self->ext.et_801D9264.unk85 > 2) {
+                        self->ext.et_801D9264.unk85 = 0;
+                        SetStep(8);
+                    }
+                } else {
+                    posX = self->posX.i.hi + g_Tilemap.scrollX.i.hi;
+                    if (self->facingLeft && posX > g_Tilemap.width - 0x80) {
+                        self->facingLeft = 0;
+                    } else if (!self->facingLeft && posX < g_Tilemap.x + 0x80) {
+                        self->facingLeft = 1;
+                    }
+                }
+            }
+        }
+        break;
+    case 7:
+        if (!AnimateEntity(&D_us_801823C4, self)) {
+            SetStep(5);
+            self->step_s = 2;
+        }
+        if (self->animFrameIdx == 8 && self->animFrameDuration == 0) {
+            newEntity = &PLAYER;
+            posX = newEntity->posX.i.hi - self->posX.i.hi;
+            posY = newEntity->posY.i.hi - self->posY.i.hi;
+            if (!self->facingLeft) {
+                posX = -posX;
+            }
+            angle = ratan2(-posY, posX);
+            if (angle > 0x280) {
+                angle = 0x280;
+            }
+            if (angle < -0x280) {
+                angle = -0x280;
+            }
+            if (GetDistanceToPlayerY() < 0x60) {
+                PlaySfxPositional(SFX_FM_EXPLODE_SWISHES);
+                newEntity = AllocEntity(&g_Entities[160], &g_Entities[192]);
+                if (newEntity != NULL) {
+                    CreateEntityFromEntity(0x44, self, newEntity);
+                    newEntity->facingLeft = self->facingLeft;
+                    if (self->facingLeft) {
+                        newEntity->posX.i.hi += 0x10;
+                    } else {
+                        newEntity->posX.i.hi -= 0x10;
+                    }
+                    newEntity->rotZ = angle;
+                    newEntity->posY.i.hi -= 0x16;
+                    newEntity->zPriority = self->zPriority + 1;
+                }
+            }
+        }
+        break;
+    case 6:
+        if (!AnimateEntity(&D_us_801823F4, self)) {
+            SetStep(5);
+        }
+        if ((GetDistanceToPlayerY() < 0x60) && (self->animCurFrame == 0x24) &&
+            (self->animFrameDuration == 0)) {
+            PlaySfxPositional(SFX_FM_EXPLODE_SWISHES);
+            for (i = 0; i < 3; i++) {
+                newEntity = AllocEntity(&D_8007A958, &D_8007A958 + 0x5E0);
```

Comment:
I didn't check the exact math but all the AllocEntity calls should be a range inside g_Entities
`g_Entities = 0x800733D8;`